rest/oidc_api_test.go

```diff
@@ -130,10 +205,162 @@ func TestGetOIDCCallbackURL(t *testing.T) {
 		locationURL, err := url.Parse(location)
 		require.NoError(t, err, "Location header should be a valid URL")
-		redirectURI := locationURL.Query().Get(authRedirectURI)
+		redirectURI := locationURL.Query().Get(requestParamRedirectURI)
 		require.NotEmpty(t, location, "redirect_uri should be available in auth URL")
 		redirectURL, err := url.Parse(redirectURI)
 		require.NoError(t, err, "redirect_uri should be a valid URL")
 		assert.Equal(t, "", redirectURL.Query().Get(authProvider))
 	})
 }
+
+func TestCallbackState(t *testing.T) {
+	authServer, err := mockAuthServer()
+	require.NoError(t, err, "Error mocking fake authorization server")
+	defer authServer.Close()
+	issuerGoogle = authServer.URL + "/google"
+	wantUsername := "google_noah"
+
+	t.Run("check whether state is maintained when callback state is disabled explicitly", func(t *testing.T) {
+		providerGoogle := auth.OIDCProvider{
+			Name: providerNameGoogle, Issuer: issuerGoogle, ClientID: clientIDGoogle, IncludeAccessToken: true,
+			UserPrefix: providerNameGoogle, ValidationKey: &validationKeyGoogle, Register: true,
+			DiscoveryURI: issuerGoogle + wellKnownPath, DisableCallbackState: base.BoolPtr(true),
+		}
+		providers := auth.OIDCProviderMap{providerGoogle.Name: &providerGoogle}
+		options := auth.OIDCOptions{Providers: providers, DefaultProvider: &providerGoogle.Name}
+		restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &options}}
+		restTester := NewRestTester(t, &restTesterConfig)
+		defer restTester.Close()
+		fakeSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+		defer fakeSyncGateway.Close()
+		mockSyncGatewayURL = fakeSyncGateway.URL
+
+		// Initiate OpenID Connect Authorization Code flow.
+		requestURL := fmt.Sprintf("%s/db/_oidc?provider=google&offline=true", mockSyncGatewayURL)
+		request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+		require.NoError(t, err, "Error creating new request")
+		response, err := http.DefaultClient.Do(request)
+		require.NoError(t, err, "Error sending request")
+		require.Equal(t, http.StatusOK, response.StatusCode)
+
+		// Validate received token response
+		var receivedToken OIDCTokenResponse
+		require.NoError(t, err, json.NewDecoder(response.Body).Decode(&receivedToken))
+		assert.NotEmpty(t, receivedToken.SessionID, "session_id doesn't exist")
+		assert.Equal(t, wantUsername, receivedToken.Username, "name mismatch")
+		assert.Equal(t, wantTokenResponse.IDToken, receivedToken.IDToken, "id_token mismatch")
+		assert.Equal(t, wantTokenResponse.RefreshToken, receivedToken.RefreshToken, "refresh_token mismatch")
+		assert.Equal(t, wantTokenResponse.AccessToken, receivedToken.AccessToken, "access_token mismatch")
+		assert.Equal(t, wantTokenResponse.TokenType, receivedToken.TokenType, "token_type mismatch")
+		assert.True(t, wantTokenResponse.Expires >= receivedToken.Expires, "expires_in mismatch")
+
+		// Query db endpoint with bearer token
+		var responseBody db.Body
```

Comment:
We should try to keep `db.Body` use only for user document bodies. It has some special handling for Marshal/Unmarshal that we probably want to avoid for generic uses.

```suggestion
		var responseBody map[string]interface{}
```

Reply:
Added changes to keep db.Body use only for user document bodies.